File: Homework1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(IReadOnlyCollection<string> args, 
        out double val1, 
        out CalculatorOperation operation, 
        out double val2)
    {
        if (!IsArgLengthSupported(args))
            throw new ArgumentException($"Need 3 arguments, was given {args.Count}");

        if (!(double.TryParse(args.ElementAt(0), out val1) && double.TryParse(args.ElementAt(2), out val2)))
            throw new ArgumentException($"Wrong arguments, expected numbers, was given {args.ElementAt(0)} and {args.ElementAt(2)}");

        operation = ParseOperation(args.ElementAt(1));
    }

    private static bool IsArgLengthSupported(IReadOnlyCollection<string> args) => args.Count == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        return arg switch
        {
            "+" => CalculatorOperation.Plus,
            "-" => CalculatorOperation.Minus,
            "*" => CalculatorOperation.Multiply,
            "/" => CalculatorOperation.Divide,
            _ => throw new InvalidOperationException($"Expected operation(+, -, *, /), was given {arg}")
        };
    }
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator
open Microsoft.FSharp.Core


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3
    
let parseFloat (arg : string) = 
    match Double.TryParse arg  with
        | true, v -> v  
        | false, _ -> ArgumentException($"Expected number, was given {arg}") |> raise
    
let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> ArgumentException($"Expected +, -, * or /, was given {arg}") |> raise
    
let parseCalcArguments(args : string[]) =
    match isArgLengthSupported args with
    | false -> ArgumentException($"Need 3 arguments, was given {args.Length}") |> raise
    | true -> { arg1 = args[0] |> parseFloat
                arg2 = args[2] |> parseFloat
                operation = args[1] |> parseOperation }
